### PR TITLE
Browser Extension 2.0

### PR DESCRIPTION
**Pull request by svenefftinge**
This is a rewrite of the browser extension using the plasma framework, which allows to use React components and has a couple of other advantages.

The extension supports
 - github
 - gitlab
 - bitbucket
 - bitbucket-server
 
 installations on any domain.
 
 It also now starts workspaces with `autostart=true` which will automatically forward when landing in gitpod's create workspace page. A secondary button to "Open with options" is available in case a user wants to select a different IDE/ws class.

The rewrite also includes tests for the element selectors used to check if and where a button should be installed, so we get noticed when a website changes its design and breaks our integration.

### CODE REVIEW

**Comment by svenefftinge** [comment]
Here's a build of the current version:

[chrome-mv3-prod.zip](https://github.com/gitpod-io/browser-extension/files/12620313/chrome-mv3-prod.zip)

File: src/button/button.tsx [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+import { useStorage } from "@plasmohq/storage/hook";
+import React from "react";
+
+export interface GitpodButtonProps {
+  application: SupportedApplication;
+  additionalClassNames?: string[];
+}
+
+export const GitpodButton = ({ application, additionalClassNames }: GitpodButtonProps) => {
+  const [address] = useStorage<string>(storage_key);
```
Comment (filiptronicek) [inline src/button/button.tsx:15]: On first install of the extension, no base URL is set, so it breaks by navigating to `undefined`. We can fix it by defaulting to gitpod.io, which we can probably extract to a constant somewhere. 
```suggestion
  const [address] = useStorage<string>(storage_key, DEFAULT_BASE_URL);
```

Comment (svenefftinge) [reply src/button/button.tsx]: 👍  I have inlined it without a constant for since we only use it in one place.

**Comment by svenefftinge** [comment]
@filiptronicek and @selfcontained I have fixed the issues we've seen yesterday. 👍

File: src/popup.tsx [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  const [address, setAddress] = useStorage<string>(storage_key, "https://gitpod.io");
+  const [error, setError] = useState<string>();
+  const updateAddress = useCallback((address: string) => {
+    address = address.trim();
+    address = address.endsWith("/") ? address.slice(0, -1) : address;
+    const err = validate(address);
+    setError(err);
+    setAddress(address);
+  }, [setAddress, setError]);
+
```
Comment (filiptronicek) [inline src/popup.tsx:23]: in a follow up, I'll try to add the [Desktop App parser](https://github.com/gitpod-io/gitpod-desktop-app/blob/main/src/utils/parseEndpoint.ts) here, so more formats are accepted.

File: src/popup.tsx
Comment (filiptronicek) [inline src/popup.tsx:1]: For the skateboard, this does the job well. Would love to match this closer to the Dashboard UI afterwards.

File: dev/http-file-sync/main.go
Comment (filiptronicek) [inline dev/http-file-sync/main.go:1]: This is really cool, love it

File: src/button/button.css
Comment (filiptronicek) [inline src/button/button.css:1]: We could look into how this looks in all of the different [GitLab themes](https://docs.gitlab.com/ee/user/profile/preferences.html), but from what I've tested, it looks great everywhere even in dark mode.

**Review by filiptronicek — APPROVED** [review]
Thanks a lot for taking this on, @svenefftinge. It lays a solid foundation upon which we can continue to build v2.

We also seem to be getting manifest version 3 support for free with this (could close https://github.com/gitpod-io/browser-extension/pull/65) as well as an easier way to publish to Safari. 

I've only tested on Chrome, because both of my Firefox installs kept saying that the packaged extension was corrupt, but I'm guessing that's something wrong on my end.

File: src/button/button.tsx [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+          e.stopPropagation();
+          toggleDropdown();
+        }}>
+          <svg width="18" viewBox="0 0 24 24" className={classNames("chevron-icon")}>
+            <path d="M7 10L12 15L17 10H7Z"></path>
+          </svg>
+        </div>
+      </div>
+
+      {showDropdown && (
```
[the remark this answers is not in the document]
Comment (svenefftinge) [reply src/button/button.tsx:65]: https://linear.app/gitpod/issue/EXP-620/make-the-buttons-keyboard-accessible

File: test/src/button-contributions.spec.ts
[the remark this answers is not in the document]
Comment (svenefftinge) [reply test/src/button-contributions.spec.ts:1]: https://linear.app/gitpod/issue/EXP-618/nightly-integration-tests

File: src/button/button.tsx [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    document.addEventListener('click', handleDocumentClick);
+    return () => {
+      document.removeEventListener('click', handleDocumentClick);
+    };
+  }, []);
+
+  return (
+    <div id="gitpod-btn-nav" title="Gitpod" className={classNames("gitpod-button", application, ...(additionalClassNames || []))}>
+      <div className={classNames("button")}>
+        <a className={classNames("button-part", "action")} href={actions[0].href}>
```
[the remark this answers is not in the document]
Comment (svenefftinge) [reply src/button/button.tsx:49]: https://linear.app/gitpod/issue/EXP-621/add-preference-to-automatically-open-in-new-tab

File: src/popup.tsx
Comment (svenefftinge) [reply src/popup.tsx:1]: https://linear.app/gitpod/issue/EXP-622/make-the-popup-pretty-using-our-brand-design

**Comment by svenefftinge** [comment]
Thank you @filiptronicek 🙏  I'll merge this so we can add some of the things you mentioned in smaller follow-ups.

File: src/contents/button.tsx
[the remark this answers is not in the document]
Comment (svenefftinge) [reply src/contents/button.tsx:1]: https://linear.app/gitpod/issue/EXP-624/investigate-performance-improvement

File: src/popup.tsx (hunk 2 above)
Comment (svenefftinge) [reply src/popup.tsx:23]: https://linear.app/gitpod/issue/EXP-625/improve-setting-the-gitpod-url

**Pull request merged**